google/cloud/storage/internal/sha256_hash.cc

```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/sha256_hash.h"
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+std::vector<std::uint8_t> Sha256Hash(std::string const& str) {
+  SHA256_CTX sha256;
+  SHA256_Init(&sha256);
+  SHA256_Update(&sha256, str.c_str(), str.size());
+
+  unsigned char hash[SHA256_DIGEST_LENGTH];
+  SHA256_Final(hash, &sha256);
+  return {hash, hash + sizeof(hash)};
```

Comment:
Type mismatches make me slightly uncomfortable, even though it's very unlikely that uint8_t is different than unsigned char. I'm not sure if the standard *guarantees* that these types are equivalent, or if it's possible for them to differ. I might consider a `static_assert(std::is_same<std::uint8_t, unsigned char>::value, "");`. I don't feel very strongly either way. Just something to consider. I'm happy with whatever decide.

Reply:
It looks like SHA256_Final() demands that we pass `unsigned char*`, so the question boils down to what `std::int8_t = unsigned char` means, and given that SHA256_Final() delivers bytes, and both values are unsigned, I think this is right.

Reply:
I *think* this is correct as-is.  This is using the constructor from an `InputIt`:

https://en.cppreference.com/w/cpp/container/vector/vector

which should accept any iterator whole `value_type` is convertible to `std::uint8_t`.

So all we need is that `unsigned char` is convertible to `std::uint8_t`. Which is always true, this is unsigned conversion, and `std::uint8_t` must be  large enough to fit an `unsigned char`: `std::uint8_t`, if it exists, must be an integer type, therefore `sizeof(std::uint8_t) >= 1` and since `sizeof(unsigned char) == 1` q.e.d.

I added some comments because subtle things like this will surprise us in the future.

Reply:
Thanks for explaining. If this is guaranteed by the standard, there's no need for the static_assert. Feel free to remove it if you want. Your call.

Reply:
I don't think it matters that "`unsigned char` fits in a `std::uint_8`", despite agreeing that (if `std::uint8_t` exists) it always does, because SHA256_Final() only delivers 8-bit, unsigned values, and they always fit in `std::uint8_t`.